### PR TITLE
fix(popup): prevent data loss in connect screen

### LINKED ISSUE #293: [FEATURE REQUEST]  - UX - temp store text entered before connecting wallet
### Describe the feature you would like to request

Before connecting a wallet to the extension, allow the extension to temporarily retain any text entered on the settings page. 

**Use-case**: if a user enters some text on the UI; navigates away to upload the extension's public key into their wallet; and returns to the extension, the user's previously entered text currently gets cleared.

### Describe the solution you would like

Build a temporary storage mechanism for the extension to retain text entered on the Settings page, even before connecting to a wallet. 

**Impact**: as long as the browser remains open, if a user enters some text on the settings pop-up, and then switches to complete another task before connecting their wallet, the extension must keep the entered text such that the text is populated on the UI when the user opens the extension. 

### Additional context

When the browser is closed (or opened for the first time), this temporary storage should be empty.

### PR DESCRIPTION
**Pull request by dianafulga**
Closes #293

### CODE REVIEW

**Comment by raducristianpopa** [comment]
<h2>Extension builds preview</h2>

<table role="table">
<thead>
  <tr>
    <th>Name</th>
    <th>Link</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td align="center">Latest commit</td>
    <td align="center">0848a39f18098997d35a909eb0fd5b305908d424</td>
  </tr>
  <tr>
    <td align="center">Latest job logs</td>
    <td align="center"><a href="https://github.com/interledger/web-monetization-extension/actions/runs/9452965342">Run #9452965342</a></td>
  </tr>
  <tr><td align="center">Badge</td><td align="center"><a href="https://github.com/interledger/web-monetization-extension/suites/24737179735/artifacts/1586514384">Download</a></td></tr><tr><td align="center">Badge</td><td align="center"><a href="https://github.com/interledger/web-monetization-extension/suites/24737179735/artifacts/1586514386">Download</a></td></tr>
</tbody>
</table>

**Comment by sidvishnoi** [comment]
While this works, I don't think this is the best approach to do this. We're storing UI data and not "extension state" here, and adding extra layers of communication.
I think for the use case of preventing data loss, we can simply store this data within ConnectWalletForm with browser's localStorage or sessionStorage (these are available in popup context). What do you think @dianafulga?

**Review by sidvishnoi — COMMENTED** [review]
While this works, I don't think this is the best approach to do this. We're storing UI data and not "extension state" here, and adding extra layers of communication.
I think for the use case of preventing data loss, we can simply store this data within ConnectWalletForm with browser's localStorage or sessionStorage (these are available in popup context). What do you think @dianafulga?

**Comment by sidvishnoi** [comment]
A quick pass with `localStorage`.. Please test, I didn't test enough. Maybe sessionStorage could work too.

```diff
diff --git a/src/popup/components/ConnectWalletForm.tsx b/src/popup/components/ConnectWalletForm.tsx
index ce37049..9a589e7 100644
--- a/src/popup/components/ConnectWalletForm.tsx
+++ b/src/popup/components/ConnectWalletForm.tsx
@@ -27,6 +27,10 @@ interface ConnectWalletFormProps {
 }
 
 export const ConnectWalletForm = ({ publicKey }: ConnectWalletFormProps) => {
+  const [formState, setFormState] = useLocalStorage<ConnectWalletFormInputs>(
+    'form_state',
+    { amount: '', recurring: false, walletAddressUrl: '' }
+  )
   const {
     register,
     handleSubmit,
@@ -39,7 +43,7 @@ export const ConnectWalletForm = ({ publicKey }: ConnectWalletFormProps) => {
     mode: 'onSubmit',
     reValidateMode: 'onBlur',
     defaultValues: {
-      recurring: false
+      ...formState
     }
   })
   const [currencySymbol, setCurrencySymbol] = React.useState<{
@@ -147,7 +151,9 @@ export const ConnectWalletForm = ({ publicKey }: ConnectWalletFormProps) => {
         {...register('walletAddressUrl', {
           required: { value: true, message: 'Wallet address URL is required.' },
           onBlur: (e: React.FocusEvent<HTMLInputElement>) => {
-            getWalletCurrency(e.currentTarget.value)
+            const walletAddressUrl = e.currentTarget.value
+            setFormState((s) => ({ ...s, walletAddressUrl }))
+            getWalletCurrency(walletAddressUrl)
           }
         })}
       />
@@ -174,10 +180,12 @@ export const ConnectWalletForm = ({ publicKey }: ConnectWalletFormProps) => {
           required: { value: true, message: 'Amount is required.' },
           valueAsNumber: false,
           onBlur: (e: React.FocusEvent<HTMLInputElement>) => {
-            setValue(
-              'amount',
-              formatNumber(+e.currentTarget.value, currencySymbol.scale)
+            const amount = formatNumber(
+              +e.currentTarget.value,
+              currencySymbol.scale
             )
+            setFormState((s) => ({ ...s, amount }))
+            setValue('amount', amount)
           }
         })}
       />
@@ -198,3 +206,25 @@ export const ConnectWalletForm = ({ publicKey }: ConnectWalletFormProps) => {
 }
 
 const HOSTS_PERMISSION = { origins: ['http://*/*', 'https://*/*'] }
+
+import { useState, useEffect } from 'react'
+
+function useLocalStorage<T>(key: string, defaultValue?: T) {
+  const [value, setValue] = useState<T>(() => {
+    if (typeof localStorage === 'undefined') {
+      return defaultValue
+    }
+    const storedValue = localStorage.getItem(key)
+    if (storedValue) {
+      return JSON.parse(storedValue)
+    }
+    return defaultValue
+  })
+
+  useEffect(() => {
+    if (value === undefined) return
+    localStorage.setItem(key, JSON.stringify(value))
+  }, [value, key])
+
+  return [value, setValue] as const
+}
```

File: src/background/services/storage.ts [hunk 1]
```diff
@@ -82,6 +81,24 @@ export class StorageService {
...
+      'amount',
+      'rateOfPay',
+      'minRateOfPay',
+      'maxRateOfPay',
+      'walletAddress',
+      'publicKey'
+    ])
+    await this.browser.storage.local.set({ ...storageData, ...data })
+  }
+
```
Comment (sidvishnoi) [inline src/background/services/storage.ts:101]: This is unrelated now, right? Can be removed?

File: src/popup/components/ConnectWalletForm.tsx [hunk 2]
```diff
@@ -105,7 +146,8 @@ export const ConnectWalletForm = ({ publicKey }: ConnectWalletFormProps) => {
           required: { value: true, message: 'Wallet address URL is required.' },
           onBlur: (e: React.FocusEvent<HTMLInputElement>) => {
             getWalletCurrency(e.currentTarget.value)
-          }
+          },
+          onChange: handleOnChangeWalletAddressUrl
```
Comment (sidvishnoi) [inline src/popup/components/ConnectWalletForm.tsx:150]: Any reason to not do it on blur? Does it not fire right before popup closes? We can avoid debounce and stuff by doing it on blur.

File: src/shared/messages.ts [hunk 3]
```diff
@@ -45,6 +45,12 @@ export interface UpdateRateOfPayPayload {
   rateOfPay: string
 }
 
+export interface SetPopupDataPayload {
+  walletAddressUrl?: string
+  amountValue?: string
+  recurring?: boolean
+}
+
```
Comment (sidvishnoi) [inline src/shared/messages.ts:53]: Can be removed.

File: src/popup/components/ConnectWalletForm.tsx [hunk 4]
```diff
@@ -37,32 +37,73 @@ export const ConnectWalletForm = ({ publicKey }: ConnectWalletFormProps) => {
     mode: 'onSubmit',
     reValidateMode: 'onBlur',
     defaultValues: {
-      recurring: false
+      recurring: localStorage?.getItem('recurring') === 'true' || false,
+      amount: localStorage?.getItem('amountValue') || undefined,
+      walletAddressUrl: localStorage?.getItem('walletAddressUrl') || undefined
```
Comment (sidvishnoi) [inline src/popup/components/ConnectWalletForm.tsx:42]: As the collective JSON object won't be too big, I guess we can store these together? Any rationale to not to that?

**Review by sidvishnoi — CHANGES_REQUESTED** [review]
Some questions and small suggestions...

Also, were there some issues in the `useLocalStorage` hook based approach I suggested? I think it could've been used as is.

File: src/background/services/storage.ts (hunk 1 above)
Comment (dianafulga) [reply src/background/services/storage.ts:101]: Currently not used, but I thought it's a good util to be used in the feature if we want.

File: src/popup/components/ConnectWalletForm.tsx (hunk 2 above)
Comment (dianafulga) [reply src/popup/components/ConnectWalletForm.tsx:150]: clicking outside the extension didn't trigger the onBlur

File: src/popup/components/ConnectWalletForm.tsx (hunk 4 above)
Comment (dianafulga) [reply src/popup/components/ConnectWalletForm.tsx:42]: seemed cleaner than reading, parsing, changing, writing

**Comment by dianafulga** [comment]
> Also, were there some issues in the `useLocalStorage` hook based approach I suggested? I think it could've been used as is.

There is no need for using/keeping state, we only use the localStorage for initialisation and writing should be async

File: src/background/services/storage.ts (hunk 1 above)
Comment (sidvishnoi) [reply src/background/services/storage.ts:101]: I tend to keep PRs as focused as they can be, but alright.

**Review by sidvishnoi — DISMISSED** [review]

**Review by sidvishnoi — APPROVED** [review]
